File: sample/src/main/java/contacts/sample/cheatsheet/basics/java/QueryContactsAdvanced.java
package contacts.sample.cheatsheet.basics.java;

import static contacts.core.WhereKt.*;
import static contacts.core.util.ContactLookupKeyKt.lookupKeyIn;

import android.accounts.Account;
import android.app.Activity;

import java.util.List;

import contacts.core.ContactsFactory;
import contacts.core.Fields;
import contacts.core.entities.Contact;

public class QueryContactsAdvanced extends Activity {

    Contact getContactById(Long contactId) {
        return ContactsFactory.create(this)
                .query()
                .where(
                        equalTo(Fields.Contact.Id, contactId)
                )
                .find()
                .get(0);
    }

    List<Contact> getContactByLookupKey(String lookupKey) {
        return ContactsFactory.create(this)
                .query()
                .where(
                        lookupKeyIn(Fields.Contact, lookupKey)
                )
                .find();
    }

    List<Contact> getAllContactsForAGoogleAccount() {
        return ContactsFactory.create(this)
                .query()
                .accounts(new Account("email@gmail.com", "com.google"))
                .find();
    }

    List<Contact> getOnlyFavoriteContacts() {
        return ContactsFactory.create(this)
                .query()
                .where(
                        equalTo(Fields.Contact.Options.Starred, true)
                )
                .find();
    }

    List<Contact> getContactsPartiallyMatchingDisplayName() {
        return ContactsFactory.create(this)
                .query()
                .where(
                        contains(Fields.Contact.DisplayNamePrimary, "alex")
                )
                .find();
    }

    List<Contact> getContactsWithAtLeastOneGmailEmail() {
        return ContactsFactory.create(this)
                .query()
                .where(
                        endsWith(Fields.Email.Address, "@gmail.com")
                )
                .find();
    }

    List<Contact> getContactsWithAtLeastOnePhoneNumber() {
        return ContactsFactory.create(this)
                .query()
                .where(
                        isNotNullOrEmpty(Fields.Phone.Number)
                        // or equalTo(Fields.Contact.HasPhoneNumber, true)
                )
                .find();
    }

    List<Contact> getContactsWithAtLeastOnePhoneNumberAndEmail() {
        return ContactsFactory.create(this)
                .query()
                .where(
                        and(
                                isNotNullOrEmpty(Fields.Phone.Number),
                                // or equalTo(Fields.Contact.HasPhoneNumber, true),
                                isNotNullOrEmpty(Fields.Email.Address)
                        )
                )
                .find();
    }
}

File: sample/src/main/java/contacts/sample/cheatsheet/basics/kotlin/QueryContactsAdvanced.kt
package contacts.sample.cheatsheet.basics.kotlin

import android.accounts.Account
import android.app.Activity
import contacts.core.*
import contacts.core.entities.Contact

class QueryContactsAdvanced : Activity() {

    fun getAllContactsForAGoogleAccount(): List<Contact> = Contacts(this)
        .query()
        .accounts(Account("email@gmail.com", "com.google"))
        .find()

    fun getOnlyFavoriteContacts(): List<Contact> = Contacts(this)
        .query()
        .where {
            Contact.Options.Starred equalTo true
        }
        .find()

    fun getContactsPartiallyMatchingDisplayName(): List<Contact> = Contacts(this)
        .query()
        .where {
            Contact.DisplayNamePrimary contains "alex"
        }
        .find()

    fun getContactsWithAtLeastOneGmailEmail(): List<Contact> = Contacts(this)
        .query()
        .where {
            Email.Address endsWith "@gmail.com"
        }
        .find()

    fun getContactsWithAtLeastOnePhoneNumber(): List<Contact> = Contacts(this)
        .query()
        .where {
            Phone.Number.isNotNullOrEmpty()
            // or Contact.HasPhoneNumber equalTo true
        }
        .find()

    fun getContactsWithAtLeastOnePhoneNumberAndEmail(): List<Contact> = Contacts(this)
        .query()
        .where {
            Phone.Number.isNotNullOrEmpty() and Email.Address.isNotNullOrEmpty()
            // or Contact.HasPhoneNumber equalTo true and Email.Address.isNotNullOrEmpty()
        }
        .find()
}